### DISCUSSION TITLE
Upload Progress (solved) &  Memory issues(solved)

### DISCUSSION THREAD

**Discussion by hstarmans** [miguelgrinberg/microdot#222] [Q&A]
Thank you for microdot and the upload [example](https://github.com/miguelgrinberg/microdot/tree/main/examples/uploads).

I have two questions:
 - how do I add a upload progress?  (issues solved)

I just checked and my modification does now work.

Fetch does not support upload progress, as such XMLHttpRequest is often suggested. 

- I have memory issues with multiple upload requests

If I sent multiple request to the upload button, microdot runs out of memory.  
  --> Memory issues are caused by https://github.com/miguelgrinberg/microdot/discussions/198 and have been solved

```javascript

  async function uploadnew(ev) {
    ev.preventDefault();
    const file = uploadformFile.files[0];
    const payload = new FormData();
    payload.append(file.name, file);
    const req = new XMLHttpRequest();
    req.open("POST", '/upload');
    req.setRequestHeader('Content-Disposition', `attachment; filename="${file.name}"`);

    // how does microdot sent back a progress event?
    req.upload.addEventListener('progress', function(e){
      console.log( e.loaded + " loaded of " + e.total );
    })

    req.send(payload);
  }
```
.

**Comment by miguelgrinberg** [comment]
> How do I add a upload progress?

This is a client-side problem. XMLHttpRequest is probably the way to go.

> I have memory issues with multiple upload requests

This can be addressed with standard synchronization primitives. If you want to support only one upload at a time you can use a `Lock`. If you want to allow "n" uploads then a `Semaphore` would handle that. If more tasks that the number configured attempt to do an upload the extra requests will wait until the currently running ones end before starting.

**Comment by hstarmans** [comment]
Thank you for your advice.
I added upload progress via XMLHttpRequest and discovered a small issue with your example.
If you cancel the upload the loop will keep existing in the background.

To clarify, this is what you have in the example

```python
# write the file to the files directory in 1K chunks
  with open('files/' + filename, 'wb') as f:
      while size > 0:
          chunk = await request.stream.read(min(size, 1024))
          f.write(chunk)
          size -= len(chunk)
```
I see a problem if the request is canceled from the client side. In my case, the stream keeps reading zero chunks.
This is what I believe should resolve the issue. You can probably write it better but you get the idea.
```python
        # write the file to the files directory in 1K chunks
        with open(folder + "/" + filename, "wb") as f:
            fail = 0
            while size > 0:
                chunk = await request.stream.read(min(size, 1024))
                f.write(chunk)
                size -= len(chunk)
                # no new chunk coming in, user hanged up
                if len(chunk) == 0:
                    fail += 1
                if fail > 10:
                    break
                print(f"processed {size}")
        if fail > 10:
            print(f"removing {folder}/{filename}")
            os.remove(folder + "/" + filename)
            return {"error": "user hanged up"}, 414
```

I have memory issues on the esp32 devkit 1.  I don't have memory issues on the ESP32 S3.
I render a utemplate which is rather large for an index page. 
I resolved it via the suggestions here (https://github.com/miguelgrinberg/microdot/discussions/198)
